ruby_programming/automated_testing/rspec_part_one_basics.md

```diff
@@ -0,0 +1,822 @@
+~~~
+
+As you have probably been anticipating, this test did indeed fail. The output is coloured red which signifies a failing test.
+
+The output also displays some useful information indicating what went wrong. There are two important parts of the output in-particular that will help you debug why it failed.
+
+The first part, is the output showing what the test expected and what it got instead. This is usually enough to let you know what went wrong and where to start with fixing the test.
+
+~~~ruby
+expected: "My List (1 task)"
+got: "My List (0 tasks)"
+~~~
+
+Another useful part of the output is the line number it failed on. In larger, more complicated test files where the expected and got output doesn't provide enough of a clue as to what went wrong. This will provide a useful place to start when tracking down why the test failed
+
+~~~ruby
+# ./spec/todo_list_spec.rb:16:in `block (3 levels) in <top (required)>'
+~~~
+
+Let's quickly fix this test to get it working. We just need to pass in a tasks array which contains one task:
+
+~~~ruby
+# spec/todo_list_spec.rb
+
+it "returns the name with 1 task remaining" do
+  todo_list = TodoList.new("My List", ["task 1"])
+
+  expect(todo_list.name).to eq("My List (1 task)")
+end
+~~~
+
+Run the tests again with `$ rspec`. We should now have 2 passing tests and one pending test remaining.
+
+### Matchers
+
+So far we have been using `eq` matcher in our expectations. But RSpec has a wide selection of different matchers we can use to make the tests more expressive in all sorts of different scenarios.
+
+We'll cover a few matchers that you are likely to find the most useful here. But we encourage you to have a browse through the [RSpec matchers documentation](https://relishapp.com/rspec/rspec-expectations/docs/built-in-matchers) to see what other matchers are at your disposal.
+
+#### The Eq Matcher
+
+The `eq` matcher is the most versatile and commonly used matcher. It will likely be the matcher you use the most. It simply compares the actual value against the expected value and will return true if it **matches**. Otherwise it will return false.
+
+Lets see this matcher in action once again by finishing off the tests for the todo list name method:
+
+~~~ruby
+# spec/todo_list_spec.rb
+
+it "returns the name with 2 tasks remaining" do
+  todo_list = TodoList.new("My List", ["task 1", "task 2"])
+
+  expect(todo_list.name).to eq("My List (2 tasks)")
+end
+~~~
+
+Run all the tests in the terminal once again with `$ rspec` to see all of our todo list name method tests now passing.
+
+#### The Be Matcher
+
+We'll often need to test predicate methods in our classes that return either true or false. We could use the `eq` matcher to test these methods.
+
+But RSpec has a more expressive `be` matcher which is better suited to the job and will make the tests much nicer to read as you can expect the result to *be true* or *be false* rather than equal to true or equal to false which doesn't quite read as well.
```

Comment:
Are we covering predicate matchers in a later lesson or just leaving it to the student to find out more about them later?

Reply:
The exercises for this lesson are going to be practicing using matchers. I was thinking about including them as part of those.

Reply:
Ahh yeah that sounds good. There are defo more important matchers to cover. I just wasn't sure what the proposed structure would be. Exercises is a great place for that.